recipes/armadillo/all/conanfile.py

```diff
@@ -0,0 +1,414 @@
+
+    def configure(self):
+        if self.options.shared:
+            del self.options.fPIC
+
+        # Use default MKL path or throw error if no default path exists
+        if self.options.use_blas == "system_mkl" and not self.options.mkl_library_path:
+            if self.settings.os == "Linux":
+                self.options.mkl_library_path = "/opt/intel/mkl/lib/intel64"
+            elif self.settings.os == "Windows":
+                self.options.mkl_library_path = (
+                    "C:/IntelSWTools/compilers_and_libraries/windows/mkl/lib/intel64"
+                )
+            else:
+                raise ConanInvalidConfiguration(
+                    "A default mkl_library_path value for your operating system is not available. Please specify a value for mkl_library_path"
+                )
+
+    def validate(self):
+        tools.check_min_cppstd(self, "11")
+
+        if self.settings.os != "Macos" and (
+            self.options.use_blas == "framework_accelerate"
+            or self.options.use_lapack == "framework_accelerate"
+        ):
+            raise ConanInvalidConfiguration(
+                "framework_accelerate can only be used on Macos"
+            )
+
+        for value, options in self.co_dependencies.items():
+            options_without_value = [
+                x for x in options if getattr(self.options, x) != value
+            ]
+            if options_without_value and (len(options) != len(options_without_value)):
+                raise ConanInvalidConfiguration(
+                    "Options {} must all be set to '{}' to use this feature. To fix this, set option {} to '{}'.".format(
+                        ", ".join(options),
+                        value,
+                        ", ".join(options_without_value),
+                        value,
+                    )
+                )
+
+        if (
+            self.options.use_lapack == "openblas"
+            and self.options.use_blas != "openblas"
+        ):
+            raise ConanInvalidConfiguration(
+                "OpenBLAS can only provide LAPACK functionality when also providing BLAS functionality. Set use_blas=openblas and try again."
+            )
+
+    def requirements(self):
+        # Optional requirements
+        openblas = "openblas/0.3.15"
+        hdf5 = "hdf5/1.12.0"
+        # atlas = "atlas/3.10.3" # Pending https://github.com/conan-io/conan-center-index/issues/6757
+        # superlu = "superlu/5.2.2" # Pending https://github.com/conan-io/conan-center-index/issues/6756
+        # arpack = "arpack/1.0" # Pending https://github.com/conan-io/conan-center-index/issues/6755
+        # flexiblas = "flexiblas/3.0.4" # Pending https://github.com/conan-io/conan-center-index/issues/6827
+
+        if self.options.use_hdf5 == "hdf5":
+            # Use the conan dependency if the system lib isn't being used
+            self.requires(hdf5)
+            self.options["hdf5"].shared = self.options.shared
```

Comment:
We usually don't accept it, because Conan Center uses static by default, it won't for armadillo and the package would be invalid. So, this option is fine.

Reply:
I'm not sure what you mean here - for a static armadillo build, `self.options["hdf5"].shared = False` - is this not consistent with the CCI defaults?

Reply:
> for a static armadillo build, self.options["hdf5"].shared = False - is this not consistent with the CCI defaults?

Yes, it's the default behavior. Usually we throw an exception instead of forcing some option to fix. But in this case, it's clear if you don't force shared option, armadillo won't work as shared library on Conan Center. Thus, this option enforcement is okay.

Reply:
Options can easily be mutated in the conan deps graph... Users could override this and break the recipes
➡️ If this is a requirements for the build or runtime then it needs to be enforced in the validate method

Reply:
I really only threw this option in because I'd seen other recipes do the same, without too much thought. I think that it's fine to remove this - some limited testing suggests that armadillo can be built as a shared lib and use hdf5 and openblas as static libs fine - these will just be built in with the shared library. The issue will come in where someone wants a static build of armadillo, and for some reason specifies one of these upstream dependencies to be shared - the entire dependency chain must be static for a static build. I'll add some validation for that part.

@uilianries I'm curious why it's obvious to you that armadillo won't work as a shared library on Conan Center without forcing hdf5 and openblas to be shared? I'm wondering if I'm missing something